### PR TITLE
Ensure validate called during tfgen

### PR DESCRIPTION
**Pull request by VenelinMartinov**
fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1852

https://github.com/pulumi/pulumi-terraform-bridge/pull/1669 removed the runtime calls to `InternalValidate` to speed up provider startup. This assumed it was called during tfgen but it turns out it was not.

This introduces an explicit call to `InternalValidate` during tfgen. Thanks to @t0yv0 for prompting the investigation and @iwahbe for helping with the muxer.

To facilitate that, I've added `InternalValidate` to the shim Provider interface. Technically breaking but seems unlikely to be depended on externally.

Calling `Validate` does not do the job as that requires passing in a valid config which is not available during tfgen.

I've added an e2e test for an sdk provider, need to add one for a pf provider.

### CODE REVIEW

File: pkg/tests/main_test.go [hunk 1]
```diff
@@ -62,31 +63,35 @@ func ensureCompiledTestProviders(wd string) error {
...
-		name        string
-		source      string
-		tfgenSource string
+		name             string
+		source           string
+		tfgenSource      string
+		expectTfgenError *string
 	}
 
+	internalErrorMsg := "Internal validation of the provider failed!"
```
Comment (iwahbe) [inline pkg/tests/main_test.go:72]:
```suggestion
	const internalErrorMsg = "Internal validation of the provider failed!"
```

File: pkg/tfgen/generate.go [hunk 2]
```diff
@@ -424,7 +424,10 @@ func (g *Generator) makePropertyType(typePath paths.TypePath,
 	case shim.TypeSet:
 		t.kind = kindSet
 	default:
-		panic("impossible: sch.Type() should be one of TypeMap, TypeList, TypeSet at this point")
+		errorStr := fmt.Sprintf(
+			"impossible: sch.Type() should be one of TypeMap, TypeList, TypeSet at this point path: %s, type: %s",
+			typePath.String(), sch.Type())
+		panic(errorStr)
```
Comment (iwahbe) [inline pkg/tfgen/generate.go:430]:
```suggestion
		contract.Failf(
			"impossible: sch.Type() should be one of TypeMap, TypeList, TypeSet at this point path: %s, type: %s",
			typePath.String(), sch.Type())
```

**Review by iwahbe — COMMENTED** [review]
This looks good. I think we will need to test on downstream providers before we can merge

File: pkg/tests/main_test.go (hunk 1 above)
Comment (VenelinMartinov) [reply pkg/tests/main_test.go:72]: this is not valid, as we can't then take the address of the constant, I've reverted it.

**Comment by VenelinMartinov** [comment]
Downstream failures:

auth0: https://github.com/pulumi/pulumi-auth0/actions/runs/8648328389/job/23711873275?pr=493 - this looks related.

**Comment by VenelinMartinov** [comment]
After fixing https://github.com/pulumi/pulumi-terraform-bridge/issues/1816 it turns out this caused quite a few downsteam failures:

```
FAILURE https://github.com/pulumi/pulumi-auth0/pull/493
FAILURE https://github.com/pulumi/pulumi-oci/pull/431
MISSING https://github.com/pulumi/pulumi-opsgenie/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-splunk/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-libvirt/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-minio/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-kafka/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-aws/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-onelogin/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-cloudamqp/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-rabbitmq/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-newrelic/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-dnsimple/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-nomad/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-mailgun/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-mysql/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-sumologic/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-vsphere/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-keycloak/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-slack/actions/workflows/upgrade-bridge.yml
MISSING https://github.com/pulumi/pulumi-vault/actions/workflows/upgrade-bridge.yml
```

Looking into it, the ones I checked don't look like actual regressions in the schema but rather an issue with the validation logic.

**Comment by VenelinMartinov** [comment]
I've kicked off another round of downstream tests, will report back here once they finish.

**Comment by t0yv0** [comment]
pf may use different mechanics for validations so not as relevant in this PR.

File: pkg/tests/main_test.go [hunk 3]
```diff
@@ -21,19 +21,20 @@ import (
...
+	"runtime"
+	"strings"
 	"testing"
 
 	"github.com/pulumi/pulumi/pkg/v3/testing/integration"
-	"runtime"
 )
 
-func TestMain(m *testing.M) {
+func TestMains(t *testing.T) {
```
Comment (t0yv0) [inline pkg/tests/main_test.go:31]: These helpers are here what usually Make is for like making sure that build targets are available before running tests. Looks like this change disables them. Is that intentional?

File: pkg/tfgen/main.go [hunk 4]
```diff
@@ -34,8 +33,12 @@ import (
 
 // Main executes the TFGen process for the given package pkg and provider prov.
 func Main(pkg string, version string, prov tfbridge.ProviderInfo) {
-	// Enable additional provider validation.
-	schema.RunProviderInternalValidation = true
+	err := prov.P.InternalValidate()
```
Comment (t0yv0) [inline pkg/tfgen/main.go:36]: Nice.

**Review by t0yv0 — APPROVED** [review]
Great!

File: pkg/tests/main_test.go (hunk 3 above)
Comment (VenelinMartinov) [reply pkg/tests/main_test.go]: Good catch, thanks! I was experimenting with these to see where they are used but didn't mean to commit. 🙌

**Pull request merged**